packages/agents-hosting/src/transcript/fileTranscriptLogger.ts

```diff
@@ -0,0 +1,409 @@
+        created: stats.birthtime
+      })
+    }
+
+    return { items, continuationToken: undefined }
+  }
+
+  /**
+   * Deletes conversation data from the store.
+   * @param channelId - The ID of the channel the conversation is in.
+   * @param conversationId - The ID of the conversation to delete.
+   * @returns A promise that represents the work queued to execute.
+   */
+  async deleteTranscript (channelId: string, conversationId: string): Promise<void> {
+    const file = this.getTranscriptFile(channelId, conversationId)
+    await this.withFileLock(file, async () => {
+      if (!await this.pathExists(file)) {
+        logger.debug(`Transcript file does not exist: ${this.protocol(file)}`)
+        return
+      }
+      await fs.unlink(file)
+    })
+  }
+
+  /**
+   * Loads a transcript from a file.
+   */
+  private async loadTranscriptAsync (transcriptFile: string): Promise<Activity[]> {
+    if (!await this.pathExists(transcriptFile)) {
+      return []
+    }
+
+    const json = await fs.readFile(transcriptFile, 'utf-8')
+    const result = JSON.parse(json)
+    return result.map(Activity.fromObject)
+  }
+
+  /**
+   * Executes a file operation with exclusive locking per file.
+   * This ensures that concurrent writes to the same transcript file are serialized.
+   */
+  private async withFileLock<T> (transcriptFile: string, operation: () => Promise<T>): Promise<T> {
+    // Get the current lock chain for this file
+    const existingLock = this._fileLocks.get(transcriptFile) ?? Promise.resolve()
+
+    // Create a new lock that waits for the existing one and then performs the operation
+    const newLock = existingLock.then(async () => {
+      return await operation()
+    }).catch(error => {
+      logger.warn('Error in write chain:', error)
+      throw error
+    })
+
+    // Update the lock chain
+    this._fileLocks.set(transcriptFile, newLock)
+
+    // Wait for this operation to complete
+    try {
+      return await newLock
+    } finally {
+      // Clean up if this was the last operation in the chain
+      if (this._fileLocks.get(transcriptFile) === newLock) {
+        this._fileLocks.delete(transcriptFile)
+      }
+    }
+  }
+
+  /**
+   * Performs the actual write operation to the transcript file.
+   */
+  private async logActivityToFile (activity: Activity, transcriptFile: string): Promise<void> {
+    const activityStr = JSON.stringify(activity)
+
+    if (!await this.pathExists(transcriptFile)) {
+      const folder = path.dirname(transcriptFile)
+      if (!await this.pathExists(folder)) {
+        await fs.mkdir(folder, { recursive: true })
+      }
+      await fs.writeFile(transcriptFile, `[${activityStr}]`, 'utf-8')
+      return
+    }
+
+    // Use file handle to append efficiently
+    const fileHandle = await fs.open(transcriptFile, 'r+')
+    try {
+      const stats = await fileHandle.stat()
+
+      // Seek to before the closing bracket
+      const position = Math.max(0, stats.size - 1)
```

Comment:
is this really better than parsing the json?

it is never a good idea to write your own parser.